precompile/assetsbridge/bridge_out.go

```diff
@@ -0,0 +1,615 @@
+	if len(recipient) == 0 {
+		return fmt.Errorf("recipient can't be empty: %v", rawInput)
+	}
+
+	switch chain {
+	case TargetChainEthereum:
+		// here we just check the length
+		if len(recipient) != 20 {
+			return fmt.Errorf("invalid recipient address format for Ethereum chain: %v", rawInput)
+		}
+	case TargetChainBitcoin:
+		_, err := btcutil.DecodeAddress(string(recipient), &chaincfg.MainNetParams)
+		if err != nil {
+			return fmt.Errorf("invalid recipient address format for Bitcoin (%v): %v", err, rawInput)
+		}
+	}
+
+	return nil
+}
+
+type TargetChain uint8
+
+const (
+	TargetChainEthereum = iota
+	TargetChainBitcoin
+)
+
+func (t TargetChain) Validate() (TargetChain, bool) {
+	if t == TargetChainEthereum || t == TargetChainBitcoin {
+		return t, true
+	}
+
+	return t, false
+}
+
+type bridgeOutInputs struct {
+	Token     common.Address
+	Amount    *big.Int
+	Chain     TargetChain
+	Recipient []byte
+}
+
+// AssetsUnlockedEventName is the name of the AssetsUnlocked event. It matches the name
+// of the event in the contract ABI.
+const AssetsUnlockedEventName = "AssetsUnlocked"
+
+// AssetsUnlockedEvent is the implementation of the AssetsUnlocked event that contains
+// the following arguments:
+// - from (indexed): the address from which the tokens are bridged out,
+// - recipient (indexed): the address to which the tokens are transferred
+// - amount (non-indexed): the amount of tokens transferred
+// - chain (non-indexed): the destination chain
+// - sequenceNumber (non-indexed): the sequenceNumber of this AssetsUnlocked
+// - token (non-indexed): the token being bridged out.
+type AssetsUnlockedEvent struct {
+	sender, token        common.Address
+	recipient            []byte
+	chain                uint8
+	unlockSequenceNumber *big.Int
+	amount               *big.Int
+}
+
+func NewAssetsUnlockedEvent(
+	from, token common.Address,
+	recipient []byte,
+	chain uint8,
+	sequenceNumber *big.Int,
+	amount *big.Int,
+) *AssetsUnlockedEvent {
+	return &AssetsUnlockedEvent{
+		sender:               from,
+		recipient:            recipient,
+		token:                token,
+		unlockSequenceNumber: sequenceNumber,
+		amount:               amount,
+		chain:                chain,
+	}
+}
+
+func (te *AssetsUnlockedEvent) EventName() string {
+	return AssetsUnlockedEventName
+}
+
+func (te *AssetsUnlockedEvent) Arguments() []*precompile.EventArgument {
+	return []*precompile.EventArgument{
+		{
+			Indexed: false,
+			Value:   te.unlockSequenceNumber,
+		},
+		{
+			Indexed: true,
+			Value:   te.recipient,
+		},
+		{
+			Indexed: true,
+			Value:   te.token,
+		},
+		{
+			Indexed: false,
+			Value:   te.sender,
+		},
+		{
+			Indexed: false,
+			Value:   te.amount,
+		},
+		{
+			Indexed: false,
+			Value:   te.chain,
+		},
+	}
+}
+
+// helpers for extracting data out of the EVM response
+
+func extractBalanceOfFromEVMResult(retData []byte) (*big.Int, error) {
+	if len(retData) != 32 {
```

Comment:
Did we confirm this is actually so simple and the return data is just a 32-byte int that can be directly used to feed a big.Int?